### PR TITLE
DAOS-7374 test: Restructure FullPoolContainerCreate with new test case

### PR DESCRIPTION
**Pull request by saurabhtandan**
Restructure FullPoolContainerCreate with new test case scenarios
as per discussions.
- Fill the pool and verify it is returning expected error message
- Verify container can be deleted in filled state
- Verify the space reclaimed after container is destroyed and aggregation is complete is close enough to the initial free space.

Signed-off-by: Saurabh Tandan <saurabh.tandan@intel.com>

### CODE REVIEW

File: src/tests/ftest/container/full_pool_container_create.py [hunk 1]
```diff
@@ -86,32 +88,56 @@ def test_no_space_cont_create(self):
...
+                        self.fail("caught unexpected exception while writing "
+                                  "object: {}".format(repr(excep)))
+                    else:
+                        # calculate the data written to second container
+                        # and verify it is under the threshold of 0.1%
+                        # of pool size
+                        written_data_cont2 = (written_data_cont2 +
+                                              (write_count * obj_sz))
+                        threshold_value = (threshold_percent *
+                                           self.pool.scm_size.value)
```
[the remark this answers is not in the document]
Comment (saurabhtandan) [reply src/tests/ftest/container/full_pool_container_create.py:129]: Sure will do. Thanks

File: src/tests/ftest/container/full_pool_container_create.py [hunk 2]
```diff
@@ -86,32 +88,56 @@ def test_no_space_cont_create(self):
...
+                        self.log.info("closing container")
+                        cont2.close()
+                        self.fail("caught unexpected exception while writing "
+                                  "object: {}".format(repr(excep)))
+                    else:
+                        # calculate the data written to second container
+                        # and verify it is under the threshold of 0.1%
+                        # of pool size
+                        written_data_cont2 = (written_data_cont2 +
+                                              (write_count * obj_sz))
```
Comment (shimizukko) [inline src/tests/ftest/container/full_pool_container_create.py:127]: I'm not clear about this calculation. This is the value that we wrote/sent to DAOS, not the actual value written to DAOS. Doesn't the outcome of the if statement on line 130 depend on our expectation or logic and not the DAOS behavior?

**Review by shimizukko — APPROVED** [review]

File: src/tests/ftest/container/full_pool_container_create.py (hunk 2 above)
Comment (saurabhtandan) [reply src/tests/ftest/container/full_pool_container_create.py:127]: The thinking behind checking the DAOS behavior is to be able to raise the alarm with enough information if anything goes bad. For example, if there is some bug and we are able to write X amount of data which is greater than the expected behaviors of 0.1% of pool size, we know how much greater it is.
If there is any other solution in your mind feel free to pitch it and we can discuss it.

Comment (shimizukko) [reply src/tests/ftest/container/full_pool_container_create.py:127]: I guess my question is about the [10, 1] obj_sz. How did you determine these? 10 might cause it to go beyond the threshold in the first iteration. For example, if the 0.1% threshold is 54, then hopefully DAOS will raise -1007 before we write the 10 six times. Then hopefully it'll raise -1007 before we write 1 five times. But what if the 0.1% threshold is 4. Then the first iteration will go beyond the threshold before giving DAOS any chance of raising -1007.

**Review by shimizukko — DISMISSED** [review]

File: src/tests/ftest/container/full_pool_container_create.py [hunk 3]
```diff
@@ -24,6 +23,9 @@ def test_no_space_cont_create(self):
         # probably should be -1007, revisit later
         err2 = "-1009"
```
Comment (ravalsam) [inline src/tests/ftest/container/full_pool_container_create.py:24]: So -1009 is still needed?

File: src/tests/ftest/container/full_pool_container_create.py [hunk 4]
```diff
@@ -86,32 +88,56 @@ def test_no_space_cont_create(self):
...
+                try:
+                    # write to second container
+                    cont2.write_an_obj(my_str, obj_sz, dkey, akey,
+                                       obj_cls="OC_SX")
+                    write_count += 1
 
-            self.log.info("opening container 2")
-            cont2.open()
+                except DaosApiError as excep:
+                    if not (err in repr(excep) or err2 in repr(excep)):
```
Comment (ravalsam) [inline src/tests/ftest/container/full_pool_container_create.py:115]: So, in this case, it supposes to fail with 1007 and not 1009, If that is the case maybe we should remove err2?

Comment (saurabhtandan) [reply src/tests/ftest/container/full_pool_container_create.py]: Thanks for pointing. I have verified we don't need to check for -1009 anymore and have removed it from the code.

File: src/tests/ftest/container/full_pool_container_create.py (hunk 3 above)
Comment (saurabhtandan) [reply src/tests/ftest/container/full_pool_container_create.py]: Not needed anymore. Removed from the code.

File: src/tests/ftest/container/full_pool_container_create.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (saurabhtandan) [reply src/tests/ftest/container/full_pool_container_create.py]: Done

**Review by ravalsam — DISMISSED** [review]

File: src/tests/ftest/container/full_pool_container_create.py [hunk 5]
```diff
@@ -86,32 +86,56 @@ def test_no_space_cont_create(self):
...
+        self.log.info("created 2nd container")
+        self.log.info("opening container 2")
+        cont2.open()
 
-            self.log.info("opening container 2")
-            cont2.open()
+        written_data_cont2 = 0
+        self.log.info("writing more objects, write expected to fail "
+                      "after writing a few Kb")
+        for obj_sz in [10, 1]:
```
Comment (rpadma2) [inline src/tests/ftest/container/full_pool_container_create.py:103]: Minor: Some comments here will give future testers to understand what is going on here. Why select object_size 10, 1? Why not a different number?

File: src/tests/ftest/container/full_pool_container_create.py [hunk 6]
```diff
@@ -86,32 +86,56 @@ def test_no_space_cont_create(self):
...
+            write_count = 0
+            threshold_value = (threshold_percent *
+                               self.pool.scm_size.value)
+            while True:
+                try:
+                    # write to second container
+                    cont2.write_an_obj(my_str, obj_sz, dkey, akey,
+                                       obj_cls="OC_SX")
```
Comment (rpadma2) [inline src/tests/ftest/container/full_pool_container_create.py:111]: Minor: Static values could be avoided and passed via YAML file. "OC_SX"

**Review by shimizukko — DISMISSED** [review]

**Review by rpadma2 — APPROVED** [review]
All my requested changes can be done on future PRs.

**Comment by sylviachanoiyee** [comment]
re-started gha workflow

**Comment by sylviachanoiyee** [comment]
We don't have a confirmation that daos guarantee  <0.1% space freed after aggregation takes place per discussion in ticket DAOS-7374. I worry this might cause intermittent failure. I think we need to revisit the original testcase.

**Comment by sylviachanoiyee** [comment]
Can we fix this testcase per Johann's comment in DAOS-7374?

**Comment by saurabhtandan** [comment]
> Can we fix this testcase per Johann's comment in DAOS-7374?

Yep the new test case implementation is on it's way. Most probably will be a skipped test as I am hitting -2001 on container destroy.

**Review by shimizukko — APPROVED** [review]

File: src/tests/ftest/container/full_pool_container_create.py [hunk 7]
```diff
@@ -34,84 +58,67 @@ def test_no_space_cont_create(self):
...
+
+        # check for free space to be returned back once aggregation is complete
+        # checking for a closer returned space value instead of exact value
+        # as the test is using scm only
+        counter = 1
+        threshold_value = free_space_before - (free_space_before *
+                                               threshold_percent)
+        free_space = self.pool.get_pool_free_space()
+        while free_space < threshold_value:
+            # try to wait for 4 x 30 secs for aggregation to be completed or
```
Comment (rpadma2) [inline src/tests/ftest/container/full_pool_container_create.py:114]: So we are waiting for two minutes? Is this acceptable in in real world scenario?
What happens if there is constant IO's happening during this time? Probably, those will get -1007. Is it not?
Assume that IO is happening, can aggregation still run during this time?

**Review by rpadma2 — COMMENTED** [review]
I m fine with the code. But, I would like to make sure are we covering real world scenario.

**Review by rpadma2 — APPROVED** [review]
LGTM.

File: src/tests/ftest/container/full_pool_container_create.py (hunk 7 above)
Comment (saurabhtandan) [reply src/tests/ftest/container/full_pool_container_create.py:114]: We are not exactly waiting for two minutes, we are waiting for maximum of up to two minutes. The first check will happen immediately, if aggregation hasn't reclaimed the space we wait for 30 secs and check again and repeat 4 more times. And if we still don't get the space back we fail the test (same thing as we are doing in other aggregation tests).
And yes aggregation doesn't return the space back immediately unless it is very small space. So even in real world scenario if somebody is playing with large amount of space they might have to wait a few seconds after removing/deleting the data.
And regarding constant IO, here we are only extracting the free space by querying, so even if it is full we should not get any error -1007 or anything else (if we do, something is broken).
Yes aggregation is always on by default and should be working in the background even if there is constant IO happening.

Comment (rpadma2) [reply src/tests/ftest/container/full_pool_container_create.py:114]: yes, I didn't mean that we are waiting for 2 minutes... We are checking every 30 secs. I am not sure how fast aggregation reclaims the space. hmm. I was thinking aggregation has low priority and should not affect (or run during) IO. I maybe wrong.

**Comment by sylviachanoiyee** [comment]
@saurabhtandan , could you document the test steps in the test plan here?
https://daosio.atlassian.net/wiki/spaces/DAOS/pages/10984751316/Container+Test+Plan

and in ticket DAOS-8400, please provide instruction at the top of ticket description on how to re-enable test when issue is resolved. Thanks!

**Pull request merged**